### PR TITLE
Conformance test plumbing: `CERTS`

### LINKED ISSUE #4463: Conformance test plumbing: `CERTS`

### PR DESCRIPTION
**Pull request by Lucsanszky**
# Description
Resolves #4463 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/CHANGELOG.md [hunk 1]
```diff
@@ -1,8 +1,12 @@
 # Version history for `cardano-ledger-conway`
 
-## 1.16.0.1
+## 1.16.1.0
 
-*
+* Added `Eq`, `Show`, `NFData` instances for `CertsEnv`
```
Comment (teodanciu) [inline eras/conway/impl/CHANGELOG.md]: `Generic` too, I believe, no?

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Certs.hs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import Cardano.Ledger.Conway (ConwayEra)
+import Cardano.Ledger.Conway.Rules
+import Cardano.Ledger.Conway.TxCert
+import Cardano.Ledger.Crypto (StandardCrypto)
+import Constrained
+import Data.Sequence (Seq)
+
+certsEnvSpec ::
+  Specification fn (CertsEnv (ConwayEra StandardCrypto))
+certsEnvSpec = TrueSpec
```
Comment (teodanciu) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Certs.hs:14]: These will get more refined in one of the next tickets , that's the idea?

File: libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/SpecTranslate/Conway/Certs.hs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+instance
+  ( SpecTranslate ctx (PParamsHKD Identity era)
+  , SpecRep (PParamsHKD Identity era) ~ Agda.PParams
+  , Inject ctx (VotingProcedures era)
+  , Inject ctx (Map (Network, Credential 'Staking (EraCrypto era)) Coin)
+  ) =>
+  SpecTranslate ctx (CertsEnv era)
+  where
+  type SpecRep (CertsEnv era) = Agda.CertEnv
```
Comment (teodanciu) [inline libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/SpecTranslate/Conway/Certs.hs:43]: Same type as for CERT rule?  It seems that the only extra field in CertsEnv is Tx, so I suppose this means that it's not relevant for the conformance test?

File: libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/ExecSpecRule/Conway/Certs.hs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  ExecSpecRule fn "CERTS" Conway
+  where
+  type ExecContext fn "CERTS" Conway = ConwayCertExecContext Conway
+
+  environmentSpec _ = certsEnvSpec
+
+  stateSpec _ _ = certStateSpec
+
+  signalSpec _ _ _ = txCertsSpec
+
```
Comment (TimSheard) [inline libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/ExecSpecRule/Conway/Certs.hs:35]: Out of curiosity, where are these things defined?
txCertsSpec is going to need some special propoerties
1) No duplicates allowed
2) Later Certs in the sequence must be aware of Certs earlier in the sequence.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Certs.hs (hunk 2 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Certs.hs:14]: I assume so

**Review by TimSheard — APPROVED** [review]
A good solid attempt getting the framework set up.
We will need to pay some attention to the 2 properties mentioned in the review in the follow-on PR

File: eras/conway/impl/CHANGELOG.md (hunk 1 above)
Comment (Lucsanszky) [reply eras/conway/impl/CHANGELOG.md]: Indeed, good catch!

File: libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/ExecSpecRule/Conway/Certs.hs (hunk 4 above)
Comment (Lucsanszky) [reply libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/ExecSpecRule/Conway/Certs.hs:35]: They are defined here: https://github.com/IntersectMBO/cardano-ledger/blob/9c94eddede4400e2d3ec926cfeec3a37bfa33b1f/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Certs.hs
Also, thanks for pointing out the above.

File: libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/SpecTranslate/Conway/Certs.hs (hunk 3 above)
Comment (Lucsanszky) [reply libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/SpecTranslate/Conway/Certs.hs:43]: Yes. In the `Agda` spec, there is only `CertEnv` and no separate env type for `CERTS`. Also, see the type of `certs-step` here: https://github.com/IntersectMBO/formal-ledger-specifications/blob/ebab02dc5fb2725dbe7e9ea830444e1b93bc2f2e/src/Ledger/Foreign/HSLedger/Cert.agda#L19

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Certs.hs (hunk 2 above)
Comment (Lucsanszky) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Certs.hs:14]: @teodanciu indeed! This PR just adds the skeleton and makes sure that everything typechecks.

**Review by teodanciu — APPROVED** [review]
Looks good to me

**Pull request merged**